### PR TITLE
[Node.js] Allow specifiying dependencies for output invokes

### PR DESCRIPTION
**Pull request by julienp**
Provider functions that take inputs as arguments, and return an output (aka output invokes), now allow specifying a `dependsOn` option. This allows programs to ensure that invokes are executed after things they depend on, similar to the [`depdendsOn` resource option](https://www.pulumi.com/docs/iac/concepts/options/dependson/).

This is based on https://github.com/pulumi/pulumi/pull/16560 which was reverted in https://github.com/pulumi/pulumi/pull/16642. Unlike the previous PR, this only adds the `dependsOn` option to output invokes (those that return an output, and use SDK invokeOutput function internally).

https://github.com/pulumi/pulumi/discussions/17710

Fixes https://github.com/pulumi/pulumi/issues/17748

### CODE REVIEW

File: sdk/nodejs/runtime/invoke.ts [hunk 1]
```diff
@@ -231,6 +232,9 @@ async function invokeAsync(
     // Wait for all values to be available, and then perform the RPC.
     const done = rpcKeepAlive();
     try {
+        // Wait for any explicit dependencies to complete before proceeding.
+        await gatherExplicitDependencies(opts.dependsOn);
```
Comment (tgummerer) [inline sdk/nodejs/runtime/invoke.ts:236]: Dumb typescript question, if only `InvokeOptions` is passed in here `opts.dependsOn` will be `undefined`? `gatherExplicitDependencies` won't do anything with that and that's how this works?

File: sdk/nodejs/tests/runtime/langhost/cases/075.invoke_depends_on/index.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        }),
+    ),
+);
+
+// By the time we serialise the arguments of the invoke, dependency.resolved
+// should be true due to dependsOn awaiting the setTimeout promise resolution
+// above. Without dependsOn, the invoke will be serialised before promise
+// resolution (since promises [microtasks] happen before timeout events
+// [macrotasks]).
+pulumi.runtime.invoke("test:index:echo", { dependency }, { dependsOn }).then((result) => {
```
Comment (tgummerer) [inline sdk/nodejs/tests/runtime/langhost/cases/075.invoke_depends_on/index.js:28]: Doesn't this need to be `invokeOutput`?  I thought this is no longer supposed to work with plain invokes? Or what am I missing?

File: sdk/nodejs/tests/runtime/langhost/cases/076.invoke_depends_on_non_resource/index.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+// This tests that invokes cannot depend on things which are not resources.
+
+let pulumi = require("../../../../../");
+
+const dependsOn = pulumi.output(Promise.resolve([Promise.resolve(1)]));
+pulumi.runtime.invoke("test:index:echo", {}, { dependsOn });
```
Comment (tgummerer) [inline sdk/nodejs/tests/runtime/langhost/cases/076.invoke_depends_on_non_resource/index.js:6]: Same question as above.

File: sdk/nodejs/tests/runtime/langhost/cases/075.invoke_depends_on/index.js (hunk 2 above)
Comment (julienp) [reply sdk/nodejs/tests/runtime/langhost/cases/075.invoke_depends_on/index.js]: You are right, this should use `invokeOutput`. This also reveals an issue that we might mistakenly pass `dependsOn` to `invoke` when not using TypeScript to intentionally bypassing the type checking.

I added a runtime check to `invoke` to guard against this and guide the user towards ${invoke}Output instead. I also added a test for this logging.

**Review by tgummerer — APPROVED** [review]

**Pull request merged**